Mathlib/Algebra/Lie/EngelSubalgebra.lean

```diff
@@ -0,0 +1,158 @@
+import Mathlib.Algebra.Lie.Weights.Basic
+import Mathlib.Data.Finset.NatAntidiagonal
+
+/-!
+# Engel subalgebras
+
+This file defines Engel subalgebras of a Lie algebra and provides basic related properties.
+
+The Engel subalgebra `LieSubalgebra.Engel R x` consists of
+all `y : L` such that `(ad R L x)^n` kills `y` for some `n`.
+
+## Main results
+
+Engel subalgebras are self-normalizing (`LieSubalgebra.normalizer_engel`),
+and minimal ones are nilpotent (TODO), hence Cartan subalgebras.
+
+* `LieSubalgebra.normalizer_eq_self_of_engel_le`:
+  Lie subalgebras containing an Engel subalgebra are self-normalizing,
+  provided the ambient Lie algebra is artinian.
+* `LieSubalgebra.isNilpotent_of_forall_le_engel`:
+  A Lie subalgebra of a noetherian Lie algebra is nilpotent
+  if it is contained in the Engel subalgebra of all its elements.
+-/
+
+open BigOperators LieAlgebra LieModule
+
+variable {R L M : Type*} [CommRing R] [LieRing L] [LieAlgebra R L]
+  [AddCommGroup M] [Module R M] [LieRingModule L M] [LieModule R L M]
+
+namespace LieSubalgebra
+
+variable (R)
+
+/-- The Engel subalgebra `Engel R x` consists of
+all `y : L` such that `(ad R L x)^n` kills `y` for some `n`.
+
+Engel subalgebras are self-normalizing (`LieSubalgebra.normalizer_engel`),
+and minimal ones are nilpotent, hence Cartan subalgebras. -/
+@[simps!]
+def Engel (x : L) : LieSubalgebra R L :=
+  { (ad R L x).maximalGeneralizedEigenspace 0 with
+    lie_mem' := by
+      simp only [AddSubsemigroup.mem_carrier, AddSubmonoid.mem_toSubsemigroup,
+        Submodule.mem_toAddSubmonoid, Module.End.mem_maximalGeneralizedEigenspace, zero_smul,
+        sub_zero, forall_exists_index]
+      intro y z m hm n hn
+      refine ⟨m + n, ?_⟩
+      rw [ad_pow_lie]
+      apply Finset.sum_eq_zero
+      intro ij hij
+      obtain (h|h) : m ≤ ij.1 ∨ n ≤ ij.2 := by rw [Finset.mem_antidiagonal] at hij; omega
+      all_goals simp [LinearMap.iterate_apply_eq_zero_of_le h, hm, hn] }
+
+lemma mem_engel_iff (x y : L) :
+    y ∈ Engel R x ↔ ∃ n : ℕ, ((ad R L x) ^ n) y = 0 :=
+  (Module.End.mem_maximalGeneralizedEigenspace _ _ _).trans <| by simp only [zero_smul, sub_zero]
+
+lemma self_mem_engel (x : L) : x ∈ Engel R x := by
+  simp only [mem_engel_iff]
+  exact ⟨1, by simp⟩
+
+/-- Engel subalgebras are self-normalizing.
+See `LieSubalgebra.normalizer_eq_self_of_engel_le` for a proof that Lie-subalgebras
+containing an Engel subalgebra are also self-normalizing,
+provided that the ambient Lie algebra is artinina. -/
+@[simp]
+lemma normalizer_engel (x : L) : normalizer (Engel R x) = Engel R x := by
+  apply le_antisymm _ (le_normalizer _)
+  intro y hy
+  rw [mem_normalizer_iff] at hy
+  specialize hy x (self_mem_engel R x)
+  rw [← lie_skew, neg_mem_iff (G := L), mem_engel_iff] at hy
+  rcases hy with ⟨n, hn⟩
+  rw [mem_engel_iff]
+  use n+1
+  rw [pow_succ', LinearMap.mul_apply]
+  exact hn
+
+variable {R}
+
+open Filter in
+/-- A Lie-subalgebra of an artinian Lie algebra is self-normalizing
```

Comment:
Some people disagree with me on this:
```suggestion
/-- A Lie-subalgebra of an Artinian Lie algebra is self-normalizing
```
so feel free to ignore. (I do note that we are capitalising `Lie` though.)

(IIRC Dijkstra joked that the supreme achievement is to have one's surname become an adjective that is so familiar that it is no longer capitalised.)

Reply:
Thanks, fixed everywhere.